synapse/app/_base.py

```diff
@@ -60,30 +72,37 @@
 logger = logging.getLogger(__name__)
 
 # list of tuples of function, args list, kwargs dict
-_sighup_callbacks = []
+_sighup_callbacks: List[
+    Tuple[Callable[..., None], Tuple[Any, ...], Dict[str, Any]]
+] = []
 
 
-def register_sighup(func, *args, **kwargs):
+def register_sighup(func: Callable[..., None], *args: Any, **kwargs: Any) -> None:
     """
     Register a function to be called when a SIGHUP occurs.
 
     Args:
-        func (function): Function to be called when sent a SIGHUP signal.
+        func: Function to be called when sent a SIGHUP signal.
         *args, **kwargs: args and kwargs to be passed to the target function.
     """
     _sighup_callbacks.append((func, args, kwargs))
 
 
-def start_worker_reactor(appname, config, run_command=reactor.run):
+def start_worker_reactor(
+    appname: str,
+    config: HomeServerConfig,
+    # mypy can't figure out what the default reactor is.
+    run_command: Callable[[], None] = reactor.run,  # type: ignore[attr-defined]
```

Comment:
There's similar comments in a bunch of places, this is due to how Twisted injects the default reactor into `twisted.internet.reactor`, mypy thinks this is a modular, not an object and gets all upset.

I don't think there's much else to do here (except maybe a mypy plugin?)

Reply:
I've come across this kind of thing in Sydent. [In one place](https://github.com/matrix-org/sydent/blob/7bdaac2774840e1f9697d30515f367c5c8c16866/sydent/sydent.py#L105) I ended up opting to introduce `SydentReactor`. I don't know if this kind of approach (with `SynapseReactor`?) would help.

If nothing obvious springs to mind, let's leave it as it is (it's more than fine!)

Reply:
Can we somehow cast `reactor` into something that satisfies the `twisted.internet.interfaces.IReactorCore` interface?

Reply:
Yes, I think `SydentReactor` was based on what we did for `SynapseReactor`. It won't help here though since we're not passing `reactor` in as a parameter, we're importing it at the top of the file. I haven't found a way to tell mypy that an import is of a different type than it hinks.

Reply:
> Can we somehow cast `reactor` into something that satisfies the `twisted.internet.interfaces.IReactorCore` interface?

I don't think you can cast on import, at least I haven't found anything that allows this.

Reply:
This would do it:

```diff
diff --git a/synapse/app/_base.py b/synapse/app/_base.py
index f2c1028b5d..0641e76bf4 100644
--- a/synapse/app/_base.py
+++ b/synapse/app/_base.py
@@ -22,13 +22,14 @@ import socket
 import sys
 import traceback
 import warnings
-from typing import TYPE_CHECKING, Awaitable, Callable, Iterable
+from typing import TYPE_CHECKING, Awaitable, Callable, Iterable, cast
 
 from cryptography.utils import CryptographyDeprecationWarning
 from typing_extensions import NoReturn
 
 import twisted
-from twisted.internet import defer, error, reactor
+from twisted.internet import defer, error, reactor as _reactor
+from twisted.internet.interfaces import IReactorCore
 from twisted.logger import LoggingFile, LogLevel
 from twisted.protocols.tls import TLSMemoryBIOFactory
 from twisted.python.threadpool import ThreadPool
@@ -57,6 +58,8 @@ from synapse.util.versionstring import get_version_string
 if TYPE_CHECKING:
     from synapse.server import HomeServer
 
+reactor = cast(IReactorCore, _reactor)
+
 logger = logging.getLogger(__name__)
 
 # list of tuples of function, args list, kwargs dict
```

Reply:
isort might yell, haven't actually linted, but that's the gist.

Reply:
> I haven't found a way to tell mypy that an import is of a different type than it thinks.

Possible ugly workaround?

```python
from twisted.internet import reactor as _reactor
reactor = cast(IReactorCore, _reactor)
```

Reply:
Yeah, that's quite ugly. Not sure what I think of it. 😢